app/controllers/application_controller.rb

```diff
@@ -1,10 +1,133 @@
     set :views, 'app/views'
+
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+  get '/' do
+    redirect "/tweets" if Helpers.is_logged_in?(session)
+    erb :index
+  end
+
+  get '/signup' do
+    redirect "tweets" if Helpers.is_logged_in?(session)
+    erb :"users/signup"
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect "/signup", locals: {message: "Please submit enter all the information to complete signup."}
+    else
+      @user = User.create(
+        username: params[:username],
+        email: params[:email],
+        password: params[:password]
+      )
+      session[:user_id] = @user.id
+      redirect "/tweets", locals: {message: "Welcome to Fwitter. Thank you for signing up."}
+    end
+  end
+
+  get '/login' do
+    redirect '/tweets'if Helpers.is_logged_in?(session)
+    erb :"users/login"
+  end
+
+  post '/login' do
+    if params[:username] == "" || params[:password] == ""
+      redirect "/login"
+    end
+
+    @user = User.find_by(username: params[:username])
+    if  @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/logout' do
+    session.clear if Helpers.is_logged_in?(session)
+    redirect "/login" if !Helpers.is_logged_in?(session)
+  end
+
+  get '/users' do
+    redirect "/login" if !Helpers.is_logged_in?(session)
+    erb :"users/index"
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :"users/show"
+  end
+
+  get '/tweets' do
+    redirect "/login" if !Helpers.is_logged_in?(session)
+    erb :"tweets/index"
   end
 
+  # Posted from /tweets/new
+  post '/tweets' do
+    redirect "/tweets/new" if params[:content].empty?
+    @user = Helpers.current_user(session)
+    @user.tweets << Tweet.create(content: params[:content])
+
+    erb :"tweets/index"
+  end
+
+  get '/tweets/new' do
+    redirect "/login" if !Helpers.is_logged_in?(session)
+    erb :"tweets/new"
+  end
+
+  get '/tweets/:id' do
+    redirect "/login" if !Helpers.is_logged_in?(session)
+    @tweet = Tweet.find(params[:id])
+    erb :"tweets/show"
+  end
+
+  # Posted from /tweets/:id/edit
+  post '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+    redirect "/tweets/#{@tweet.id}/edit" if params[:content].empty?
+    @tweet.update(content: params[:content])
+      
+    erb :"tweets/show"
+  end
+
+  get '/tweets/:id/edit' do
+    if Helpers.is_logged_in?(session)
+      @tweet = Tweet.find(params[:id])
+      if @tweet.user_id == Helpers.current_user(session).id
+        erb :"tweets/edit"
+      else
+        redirect "/tweets"
+      end
+    else
+      redirect "/login" 
+    end  
+  end
+
+  post '/tweets/:id/delete' do
+    if Helpers.is_logged_in?(session)
+      @tweet = Tweet.find(params[:id])
+      if @tweet.user_id == Helpers.current_user(session).id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.